Request exceptions should be observable in the Java RPC client

If an invalid request is send through the gateway an exception is throw. This exception should contain the reason for the failure. At the moment an exception without context is thrown.

```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: UNKNOWN

	at io.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:57)
	at io.zeebe.client.job.CreateJobTest.shouldNotPublishMessageWithNoJsonObjectAsPayload(CreateJobTest.java:163)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: UNKNOWN
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at io.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:55)
	... 27 more
Caused by: io.grpc.StatusRuntimeException: UNKNOWN
	at io.grpc.Status.asRuntimeException(Status.java:526)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:434)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:678)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:403)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/clients/java/src/main/java/io/zeebe/client/impl/TopologyRequestImpl.java
+++ b/clients/java/src/main/java/io/zeebe/client/impl/TopologyRequestImpl.java
@@ -35,7 +35,7 @@ public class TopologyRequestImpl implements TopologyRequestStep1 {
   public ZeebeFuture<Topology> send() {
     final HealthRequest request = HealthRequest.getDefaultInstance();
 
-    final ZeebeClientFutureImpl<Topology, HealthResponse, ?> future =
+    final ZeebeClientFutureImpl<Topology, HealthResponse> future =
         new ZeebeClientFutureImpl<>(TopologyImpl::new);
 
     asyncStub.health(request, future);

--- a/clients/java/src/main/java/io/zeebe/client/impl/ZeebeClientFutureImpl.java
+++ b/clients/java/src/main/java/io/zeebe/client/impl/ZeebeClientFutureImpl.java
@@ -16,50 +16,50 @@
 
 package io.zeebe.client.impl;
 
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
 import io.grpc.stub.StreamObserver;
 import io.zeebe.client.api.ZeebeFuture;
+import io.zeebe.client.cmd.ClientException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
-import java.util.function.BiFunction;
 import java.util.function.Function;
 
-public class ZeebeClientFutureImpl<T, R, P> extends CompletableFuture<T>
-    implements ZeebeFuture<T>, StreamObserver<R> {
+public class ZeebeClientFutureImpl<ClientResponse, BrokerResponse>
+    extends CompletableFuture<ClientResponse>
+    implements ZeebeFuture<ClientResponse>, StreamObserver<BrokerResponse> {
 
-  private final Function<R, T> responseMapper;
-
-  private final BiFunction<R, P, T> responsePayloadMapper;
-  private final P payload;
+  private final Function<BrokerResponse, ClientResponse> responseMapper;
 
   public ZeebeClientFutureImpl() {
-    this(r -> null);
+    this(brokerResponse -> null);
   }
 
-  public ZeebeClientFutureImpl(final Function<R, T> responseMapper) {
+  public ZeebeClientFutureImpl(final Function<BrokerResponse, ClientResponse> responseMapper) {
     this.responseMapper = responseMapper;
-    this.responsePayloadMapper = null;
-    this.payload = null;
-  }
-
-  public ZeebeClientFutureImpl(final BiFunction<R, P, T> responsePayloadMapper, P payload) {
-    this.responseMapper = null;
-    this.responsePayloadMapper = responsePayloadMapper;
-    this.payload = payload;
   }
 
   @Override
-  public T join() {
+  public ClientResponse join() {
     try {
       return get();
-    } catch (final InterruptedException | ExecutionException e) {
-      throw new RuntimeException(e);
+    } catch (final ExecutionException e) {
+      final Throwable cause = e.getCause();
+      if (cause instanceof StatusRuntimeException) {
+        final Status status = ((StatusRuntimeException) cause).getStatus();
+        throw new ClientException(status.getDescription());
+      } else {
+        throw new ClientException(cause.getMessage(), cause);
+      }
+    } catch (final InterruptedException e) {
+      throw new ClientException("Failed to receive response", e);
     }
   }
 
   @Override
-  public T join(final long timeout, final TimeUnit unit) {
+  public ClientResponse join(final long timeout, final TimeUnit unit) {
     try {
       return get(timeout, unit);
     } catch (final InterruptedException | ExecutionException | TimeoutException e) {
@@ -68,14 +68,9 @@ public class ZeebeClientFutureImpl<T, R, P> extends CompletableFuture<T>
   }
 
   @Override
-  public void onNext(final R r) {
+  public void onNext(final BrokerResponse brokerResponse) {
     try {
-      if (this.responseMapper != null) {
-        complete(responseMapper.apply(r));
-      } else {
-        complete(responsePayloadMapper.apply(r, payload));
-      }
-
+      complete(responseMapper.apply(brokerResponse));
     } catch (final Exception e) {
       completeExceptionally(e);
     }

--- a/clients/java/src/main/java/io/zeebe/client/impl/workflow/DeployWorkflowCommandImpl.java
+++ b/clients/java/src/main/java/io/zeebe/client/impl/workflow/DeployWorkflowCommandImpl.java
@@ -96,7 +96,7 @@ public class DeployWorkflowCommandImpl
       final String classpathResource) {
     ensureNotNull("classpath resource", classpathResource);
 
-    try (final InputStream resourceStream =
+    try (InputStream resourceStream =
         getClass().getClassLoader().getResourceAsStream(classpathResource)) {
       if (resourceStream != null) {
         return addResourceStream(resourceStream, classpathResource);
@@ -149,7 +149,7 @@ public class DeployWorkflowCommandImpl
 
   @Override
   public ZeebeFuture<DeploymentEvent> send() {
-    final ZeebeClientFutureImpl<DeploymentEvent, DeployWorkflowResponse, Void> future =
+    final ZeebeClientFutureImpl<DeploymentEvent, DeployWorkflowResponse> future =
         new ZeebeClientFutureImpl<>(DeploymentEventImpl::new);
 
     asyncStub.deployWorkflow(request.build(), future);

--- a/clients/java/src/main/java/io/zeebe/client/impl/workflow/PublishMessageCommandImpl.java
+++ b/clients/java/src/main/java/io/zeebe/client/impl/workflow/PublishMessageCommandImpl.java
@@ -109,7 +109,7 @@ public class PublishMessageCommandImpl
 
   @Override
   public ZeebeFuture<Void> send() {
-    final ZeebeClientFutureImpl<Void, Empty, Void> future = new ZeebeClientFutureImpl<>();
+    final ZeebeClientFutureImpl<Void, Empty> future = new ZeebeClientFutureImpl<>();
 
     asyncStub.publishMessage(builder.build(), future);
     return future;

--- a/clients/java/src/test/java/io/zeebe/client/workflow/PublishMessageTest.java
+++ b/clients/java/src/test/java/io/zeebe/client/workflow/PublishMessageTest.java
@@ -21,6 +21,7 @@ import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.assertj.core.api.Assertions.entry;
 
 import io.zeebe.client.ZeebeClient;
+import io.zeebe.client.cmd.ClientException;
 import io.zeebe.client.util.TestEnvironmentRule;
 import io.zeebe.exporter.record.Record;
 import io.zeebe.exporter.record.value.MessageRecordValue;
@@ -161,15 +162,17 @@ public class PublishMessageTest {
   @Test
   public void shouldNotPublishMessageWithNoJsonObjectAsPayload() {
     assertThatThrownBy(
-        () ->
-            client
-                .workflowClient()
-                .newPublishMessageCommand()
-                .messageName("name")
-                .correlationKey("key")
-                .payload("[]")
-                .send()
-                .join());
+            () ->
+                client
+                    .workflowClient()
+                    .newPublishMessageCommand()
+                    .messageName("name")
+                    .correlationKey("key")
+                    .payload("[]")
+                    .send()
+                    .join())
+        .isInstanceOf(ClientException.class)
+        .hasMessageContaining("Could not read property 'payload'");
   }
 
   public static class Payload {

--- a/gateway/src/main/java/io/zeebe/gateway/RequestActor.java
+++ b/gateway/src/main/java/io/zeebe/gateway/RequestActor.java
@@ -15,9 +15,12 @@
  */
 package io.zeebe.gateway;
 
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
 import io.grpc.stub.StreamObserver;
 import io.zeebe.util.sched.Actor;
 import io.zeebe.util.sched.future.ActorFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.function.Function;
 
 public class RequestActor extends Actor {
@@ -31,17 +34,29 @@ public class RequestActor extends Actor {
             actor.runOnCompletion(
                 responseFuture,
                 (v, t) -> {
-                  try {
-                    if (t == null) {
+                  if (t == null) {
+                    try {
                       final R response = responseMapper.apply(v);
                       streamObserver.onNext(response);
                       streamObserver.onCompleted();
-                    } else {
-                      streamObserver.onError(t);
+                    } catch (Exception e) {
+                      streamObserver.onError(convertThrowable(e));
                     }
-                  } catch (final Exception e) {
-                    streamObserver.onError(e);
+                  } else {
+                    streamObserver.onError(convertThrowable(t));
                   }
                 }));
+  }
+
+  private StatusRuntimeException convertThrowable(final Throwable cause) {
+    final String description;
+
+    if (cause instanceof ExecutionException) {
+      description = cause.getCause().getMessage();
+    } else {
+      description = cause.getMessage();
+    }
+
+    return Status.INTERNAL.augmentDescription(description).asRuntimeException();
   }
 }

--- a/gateway/src/test/java/io/zeebe/gateway/util/RecordingStreamObserver.java
+++ b/gateway/src/test/java/io/zeebe/gateway/util/RecordingStreamObserver.java
@@ -17,9 +17,12 @@ package io.zeebe.gateway.util;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import io.grpc.Status;
 import io.grpc.stub.StreamObserver;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class RecordingStreamObserver<T> implements StreamObserver<T> {
 
@@ -60,6 +63,16 @@ public class RecordingStreamObserver<T> implements StreamObserver<T> {
   }
 
   public void assertErrors(Throwable... errors) {
-    assertThat(this.errors).containsOnly(errors);
+    final List<String> statusErrors =
+        Arrays.stream(errors)
+            .map(
+                e ->
+                    Status.INTERNAL
+                        .augmentDescription(e.getMessage())
+                        .asRuntimeException()
+                        .getMessage())
+            .collect(Collectors.toList());
+
+    assertThat(this.errors).extracting(Throwable::getMessage).containsOnlyElementsOf(statusErrors);
   }
 }

